_posts/2018-07-10-Play Framework-in-Scala–Action-Composition.md

```diff
@@ -0,0 +1,256 @@
+def action2(): Action[_]  = LoggingActionBuilder { request =>
+  val itemToBuy = request.body.\("itemToBuy").validate[String]
+  itemToBuy.fold({
+    _ => BadRequest("No valid item to buy on the body of request.")
+  },{ item =>
+    Ok(s"You are buying $item")
+  })
+}
+```
+
+<strong>About them, we can say:</strong>
+1. They are Actions.
+2. They are different Actions, i.e., they do different things in response to a given request.
+3. They are built from the LoggingActionBuilder, via providing the block function which is referred to on the invokeBlock method of the LoggingActionBuilder.
+4. Most importantly, both action1() and action2(), when called, will have their request logged.
+
+<h4> Explanation of the syntax </h4>
+
+Creating an Action from an ActionBuilder consists on calling the apply method of the ActionBuilder. 
+
+```scala
+trait ActionBuilder[+R[_], B] {
+  final def apply(block: R[B] => Result): Action[B] = ...
+  ...
+}
+```
+
+This apply method takes as argument the block function which "materializes" the Action. On the examples above, the curl parenthesis are a call to this apply method, and the function inside is the specific block of the Action at hand.
+
+<h2> Changing the type parameter R[_] </h2>
+
+The left type parameter of an ActionBuilder constrains the signature of the block function in the invokeBlock method:
+
+```scala
+trait ActionBuilder[+R[_], B] extends ActionFunction[Request, R] {
+  def invokeBlock[A](request: Request[A],
+                     block: R[A] => Future[Result]): Future[Result]
+```
+
+On the custom LoggingActionBuilder we used type Request, where we log a request and then apply the block function to that same request object:
+
+```scala
+def invokeBlock[A](request: Request[A],
+                   block: Request[A] => Future[Result]): Future[Result] = {
+  Logger.debug(s"Just received the following request: $request")
+  block(request)
+}
+```
+
+Often it is convenient that a custom ActionBuilder, via its invokeBlock method, 
+takes a request, does something with it, builds an instance of another type, and <strong>only then</strong> applies the block function to that new instance.
+Authentication of requests is a good example. Here, a particular request, if authenticated, is normally associated with a User. It is therefore useful to obtain the corresponding instance of User and pass it along, meaning, applying the block function to that instance.
+This is useful because now the Actions that are build through this ActionBuilder may count on the existence of an instance of User instead of a raw request. 
+This of course changes the signature of the block function that must be provided for the posterior creation of each Action.
+A naive AuthenticationActionBuilder could be:
+
+```scala
+class AuthenticatedAction[B](val parser: BodyParser[B])
+                            (implicit val executionContext: ExecutionContext) 
+                            extends ActionBuilder[AuthenticatedUser, B] {
```

Comment:
AuthenticatedUser might extend play's wrappedRequest class (if that still exists). But good stuff anyway.